File: Java/Euler (Java)/src/problems/P026xP050/P039.java
package problems.P026xP050;

import base.Problem;

/**
 * P039<br>
 * If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.
 * {20,48,52}, {24,45,51}, {30,40,50}
 * For which value of p <= 1000, is the number of solutions maximized?
 * Ans: 840
 * @return
 */
public class P039 extends Problem
{
   public P039()
   {
      problemNumber = 39;
      variations = 2;
   }

   @Override
   protected String execute(int variation)
   {
      switch (variation)
      {
         case 0:
            return bruteForce() + "";
         case -1:
         case 1:
            return faster() + "";
      }
      return null;
   }

   /**
    * 
    * @return
    */
   private int bruteForce()
   {
   	int max = 0;
   	int p = 0;
   	for (int ii = 3; ii <= 1000; ii++)
   	{
   		int count = 0;
   		for (int a = 1; a <= ii-2; a++)
   		{
   			for (int b = 1; b <= ii-a-1; b++)
   			{
   				int c = ii - a - b;
   				if (a*a+b*b == c*c)
   				{
   					// this is a solution!
   					count++;
   				}
   			}
   		}
   		if (count > max)
   		{
   			p = ii;
   			max = count;
   		}
   	}
   	return p;
   }
   
   /**
    * Check only values that /can/ solve the a^2 + b^2 = c^2 equation. Also, for the perimeter,
    * we know that if a and b are even, then c will be even and p will be even. If a and b are
    * odd, then c will be even and p will be even, if one is even and the other is odd, c will 
    * be odd and p will be even. So we only need to check even numbers of p.
    * @return
    */
   private long faster()
   {
   	int max = 0;
		int maxnum = 0;
		for (int size = 4; size <= 1000; size+=2)
		{
			int count = 0;
			for (int a = 1; a <= size - 2; a++)
			{
				// we can increase b by an amount related to c ...
				// a^2 + b^2 = c^2
				// a + b + c = p
				// a^2 + b^2 = (p - b - a)^2
				// a^2 + b^2 = p^2 - pb - pa - pb + b^2 + ba - pa + ab + a^2
				// 0 = p^2 - pb - pa + 2ba
				// (2pa - p^2) / (2a - 2p) = b;
				int b = (2*size*a - size*size) / (2*a - 2*size);
				int c = size - a - b;
				if (b < 1 || c < 1)
					continue;
				int left = a*a + b*b;
				int right = c*c;
				if (left == right)
				{
					count++;
				}
			}
			if (count > max)
			{
				max = count;
				maxnum = size;
			}
		}
		return maxnum;
   }
}

File: OpenCL/Euler (OpenCL)/src/problems/P031xP040/P039.cpp
#include "../../base/problem.h"

#include <set>

class P039: public Problem
{
	/**
	 * P039<br>
	 * If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.
	 * {20,48,52}, {24,45,51}, {30,40,50}
	 * For which value of p <= 1000, is the number of solutions maximized?
	 * Ans: 840
	 */
	public:
		P039()
		{
			problem_number = 39;
			variations = 1;
		}
		~P039() {}

		bool setup(int variation)
		{
			bool retval = false;
			cl = new CLKernel(kernelsize);
			switch (variation)
			{
				case 0:
				case -1:
					retval = brute_setup();
					break;
				case 1:
					retval = opt_setup();
					break;
				default:
					retval = false;
			}
			return retval;
		}

		bool teardown(int variation)
		{
			bool retval = false;
			switch (variation)
			{
				case 0:
				case -1:
					retval = brute_teardown();
					break;
				case 1:
					retval =opt_teardown();
					break;
				default:
					retval = false;
			}
			cl->releaseKernel();
			return retval;
		}

		void execute(int variation)
		{
			switch (variation)
			{
				case 0:
				case -1:
					brute();
					break;
				case 1:
					optimized();
					break;
				default:
					std::cout << "NOOP!" << std::endl;
					break;
			}
		}

	private:

		int size;

		/**
		 * Brute force version
		 *
		 */
		bool brute_setup()
		{
			size = 1000000;
			int initialVal[1000] = { 0 };

			cl->init("src/problems/P031xP040/P039.cl","brute");
         cl->createReadWriteMemory( sizeof(int)*1000,&answer_mem);
         cl->transferCPUObjectToGPU(sizeof(int)*1000,&answer_mem,&initialVal);
         cl->setupOpenCLKernelArgs(2,
         		sizeof(cl_mem), &answer_mem,
					sizeof(int), &size);
			cl->waitForQueue(); // sync
			setThreads(size);
			return true;
		}

		bool brute_teardown()
		{
			int ans[1000];
			if (!cl->transferGPUObjectToCPU(sizeof(int)*1000, &answer_mem, ans))
			{
				std::cout << "ERROR:: Could not read answer from GPU!" << std::endl;
				return false;
			}
			int max = 0;
			int sol = 0;
			for (int ii = 0; ii < 1000; ii++)
			{
				if (ans[ii] > max)
				{
					sol = ii;
					max = ans[ii];
				}
			}
			answer = to_string(sol);
			clReleaseMemObject(answer_mem);
			return true;
		}

		int brute()
		{
			cl->launchKernel(threadsize);
			cl->waitForQueue();
			return 0;
		}

		/**
		 *
		 */
		bool opt_setup()
		{
			size = 71;
			int initialVal = 0;

			cl->init("src/problems/P031xP040/P039.cl","optimized");
			cl->createReadWriteMemory( sizeof(int),&answer_mem);
			cl->transferCPUObjectToGPU(sizeof(int),&answer_mem,&initialVal);
			cl->setupOpenCLKernelArgs(2,
					sizeof(cl_mem), &answer_mem,
					sizeof(int), &size);
			cl->waitForQueue(); // sync
			setThreads(size);
			return true;
			return true;
		}

		bool opt_teardown()
		{
			int ans;
			if (!cl->transferGPUObjectToCPU(sizeof(int), &answer_mem, &ans))
			{
				std::cout << "ERROR:: Could not read answer from GPU!" << std::endl;
				return false;
			}

			answer = to_string(ans);
			clReleaseMemObject(answer_mem);
			return true;
		}

		int optimized()
		{
			cl->launchKernel(threadsize);
			cl->waitForQueue();
			return 0;
		}
};
